README.rst
This plugin adds task ``xitrumPackage`` to your `SBT <http://www.scala-sbt.org/>`_
project to collect all dependency .jar files for standalone Scala programs.

Compared to
`one-jar <http://www.scala-sbt.org/release/docs/Community/Community-Plugins#one-jar-plugins>`_
solutions, xitrum-package is faster. It doesn't merge the .jar files together,
the .jar files are left "as is".

xitrum-package is used in `Scala web framework Xitrum <http://xitrum-framework.github.io/>`_.

Alternatively, you can use `sbt-pack <https://github.com/xerial/sbt-pack>`_.

Usage
-----

Suppose your project looks like this:

::

  build.sbt

  project/
    plugins.sbt

  src/
    ...

  dirToCopy/      <-- Directory you want to copy to the packaged directory
    file1
    file2

  fileToCopy      <-- File you want to copy to the packaged directory

Add xitrum-package to your project
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Add to ``project/plugins.sbt``:

::

  addSbtPlugin("tv.cntt" % "xitrum-package" % "1.9")

Add to build.sbt:

::

  XitrumPackage.copy("dirToCopy", "fileToCopy")

Do the packaging
~~~~~~~~~~~~~~~~

Run:

::

  sbt xitrumPackage

All dependency .jar files and .jar files generated from your project will be
copied to directory ``target/xitrum``:

::

  target/
    xitrum/
      lib/        <-- Dependency .jar files are collected here
        dep1.jar
        dep2.jar
        yourProject.jar

      dirToCopy/  <-- The specified directory is copied here
        file1
        file2

      fileToCopy  <-- The specified file is copied here

Note that even when you don't need to copy anything, you have to write in
build.sbt:

::

  XitrumPackage.copy()

Multimodule project
-------------------

If your SBT project has
`many modules (subprojects) <http://www.scala-sbt.org/0.13.5/docs/Getting-Started/Multi-Project.html>`_
and you want to add ``xitrum-package`` to only several of them, you can use ``XitrumPackage.skip``:

::

  // Shared settings used by many modules
  val sharedSettings = ...

  // For the default root project
  override lazy val settings = super.settings ++ XitrumPackage.skip

  // If you want to skip this module
  lazy val module1 = Project(
    id = "module1",
    base = file("module1"),
    settings = sharedSettings ++ Seq(
      name := "module1"
    ) ++ XitrumPackage.skip
  )

  // If you want to xitrum-package this module
  lazy val app = Project(
    id = "main-app",
    base = file("main-app"),
    settings = sharedSettings ++ Seq(
      name := "main-app"
    ) ++ XitrumPackage.copy("bin", "config", "public")
  ).dependsOn(module1)

Example: https://github.com/xitrum-framework/xitrum-multimodule-demo

Boot script
-----------

You should create a shell script to start your Scala program like this:

* `runner.sh <https://github.com/xitrum-framework/xitrum-new/blob/master/script/runner>`_ (for *nix)
* `runner.bat <https://github.com/xitrum-framework/xitrum-new/blob/master/script/runner.bat>`_ (for Windows)

Run example:

::

  runner.sh mypackage.MyMainClass


CHANGELOG.rst
2.0.0:

* `#13 <https://github.com/xitrum-framework/xitrum-package/issues/13>`_
  Support SBT 1.x

1.9:

* `#12 <https://github.com/xitrum-framework/xitrum-package/issues/12>`_
  Copy: Skip symlink that points to its ancestor directory

1.8:

* `#11 <https://github.com/xitrum-framework/xitrum-package/issues/11>`_
  Improve copy for multimodule
* `#10 <https://github.com/xitrum-framework/xitrum-package/issues/10>`_
  Support multimodule projects

1.6:

* `#6 <https://github.com/xitrum-framework/xitrum-package/issues/6>`_
  Add skip option
* `#5 <https://github.com/xitrum-framework/xitrum-package/issues/5>`_
  Preserve executable permission of copied files

1.4:

* `#3 <https://github.com/xitrum-framework/xitrum-package/issues/3>`_
  Delete target/xitrum directory at each run


src/main/scala/XitrumPlugin.scala
import java.io.File

import sbt._
import Keys._

// See https://github.com/sbt/sbt/issues/944
// Add option preserveExecutable to IO.copyFile and IO.copyDirectory
object PreserveExecutableIO {
  // http://stackoverflow.com/questions/5368724
  def copy(source: File, target: File) {
    if (source.isDirectory) {
      if (!target.exists) target.mkdir()

      source.list.foreach { child =>
        copy(new File(source, child), new File(target, child))
      }
    } else {
      IO.copyFile(source, target)
      if (source.canExecute) target.setExecutable(true)
    }
  }
}

object XitrumPlugin extends Plugin {
  // Must be lazy to avoid null error
  // xitrumPackageNeedsPackageBin must be after xitrumPackageTask
  override lazy val settings = Seq(xitrumPackageTask, xitrumPackageNeedsPackageBin)

  val skipKey = SettingKey[Boolean]("xitrum-skip", "Do not package the current project (useful when you use SBT multiproject feature)")

  val copiesKey = SettingKey[Seq[String]]("xitrum-copies", "List of files and directories to copy")

  val skip = Seq(skipKey := true, copiesKey := Seq())

  def copy(fileNames: String*) = Seq(skipKey := false, copiesKey := fileNames)

  //----------------------------------------------------------------------------

  val xitrumPackageKey = TaskKey[Unit]("xitrum-package", "Packages to target/xitrum directory, ready for deploying to production server")

  // Must be lazy to avoid null error
  lazy val xitrumPackageTask = xitrumPackageKey <<=
    // dependencyClasspath: both internalDependencyClasspath and externalDependencyClasspath
    // internalDependencyClasspath ex: classes directories
    // externalDependencyClasspath ex: .jar files
    (dependencyClasspath in Runtime, baseDirectory, target,    crossTarget,  skipKey, copiesKey) map {
    (libs,                           baseDir,       targetDir, jarOutputDir, skip,    copyFileNames) =>
    try {
      if (!skip) doPackage(libs, baseDir, targetDir, jarOutputDir, copyFileNames)
    } catch {
      case e: Exception =>
        println("xitrum-package failed, reason:")
        e.printStackTrace()
    }
  }

  val xitrumPackageNeedsPackageBin = xitrumPackageKey <<= xitrumPackageKey.dependsOn(packageBin in Compile)

  //----------------------------------------------------------------------------

  private def doPackage(libs: Seq[Attributed[File]], baseDir: File, targetDir: File, jarOutputDir: File, copyFileNames: Seq[String]) {
    val packageDir = targetDir / "xitrum"
    deleteFileOrDirectory(packageDir)
    packageDir.mkdirs()

    // Copy dependencies to lib directory
    val libDir = packageDir / "lib"
    libs.foreach { lib =>
      val file = lib.data

      if (file.exists) {
        if (file.isDirectory) {
          // This dependency may be "classes" directory from SBT multimodule (multiproject)
          // http://www.scala-sbt.org/0.13.0/docs/Getting-Started/Multi-Project.html
          //
          // Ex:
          // /Users/ngoc/src/xitrum-multimodule-demo/module1/target/scala-2.10/classes
          // /Users/ngoc/src/xitrum-multimodule-demo/module1/target/scala-2.10/xitrum-multimodule-demo-module1_2.10-1.0-SNAPSHOT.jar
          if (file.name == "classes") {
            val upperDir = file / ".."
            (upperDir * "*.jar").get.foreach { f =>
              val fname = f.name
              if (!fname.endsWith("-sources.jar") && !fname.endsWith("-javadoc.jar"))
                IO.copyFile(f, libDir / f.name)
            }
          }
        } else {
          IO.copyFile(file, libDir / file.name)
        }
      }
    }

    // Copy .jar files created after running "sbt package" to lib directory
    // (see xitrumPackageNeedsPackageBin)
    (jarOutputDir * "*.jar").get.foreach { f => IO.copyFile(f, libDir / f.name) }

    copyFileNames.foreach { fName => doCopy(fName, baseDir, packageDir) }

    println("Packaged to " + packageDir)
  }

  private def deleteFileOrDirectory(file: File) {
    if (file.isDirectory) {
      val files = file.listFiles
      if (files != null) files.foreach { f => deleteFileOrDirectory(f) }
    }
    file.delete()
  }

  private def doCopy(fileName: String, baseDir: File, packageDir: File) {
    val from = baseDir / fileName
    if (!from.exists) return

    val to = packageDir / fileName
    PreserveExecutableIO.copy(from, to)
  }
}


src/main/scala/XitrumPackage.scala
import sbt._
import Keys._

object CopyUtil {
  def copy(source: File, target: File) {
    // http://stackoverflow.com/questions/5368724
    if (source.isDirectory) {
      // Skip if source is a symlink that points to its ancestor directory:
      // https://github.com/xitrum-framework/xitrum-package/issues/12
      val abs    = source.getAbsolutePath
      val canAbs = source.getCanonicalFile.getCanonicalPath
      if (abs != canAbs && abs.startsWith(canAbs)) {
        println(s"Skip: $abs -> $canAbs")
        return
      }

      if (!target.exists) target.mkdir()

      source.list.foreach { child =>
        copy(new File(source, child), new File(target, child))
      }
    } else {
      IO.copyFile(source, target)
    }
  }
}

object XitrumPackage extends AutoPlugin {
  override def trigger = allRequirements

  // - Must be lazy to avoid null error
  // - xitrumPackageNeedsPackageBin must be after xitrumPackageTask
  override lazy val projectSettings = Seq(xitrumPackageTask, xitrumPackageNeedsPackageBin, projectJarsTask)

  val skipKey = SettingKey[Boolean](
    "xitrumSkip",
     "Do not package the current project (useful when you use SBT multiproject feature)"
  )

  val copiesKey = SettingKey[Seq[String]](
    "xitrumCopies",
    "List of files and directories to copy"
  )

  val skip = Seq(skipKey := true, copiesKey := Seq())

  def copy(fileNames: String*) = Seq(skipKey := false, copiesKey := fileNames)

  //----------------------------------------------------------------------------

  val xitrumPackageKey = TaskKey[Unit](
    "xitrumPackage",
    "Packages to target/xitrum directory, ready for deploying to production server",
    KeyRanks.ATask  // This task is listed with command "sbt tasks"
  )

  // Must be lazy to avoid null error
  lazy val xitrumPackageTask = xitrumPackageKey := {
    // dependencyClasspath: both internalDependencyClasspath and externalDependencyClasspath
    // internalDependencyClasspath ex: classes directories
    // externalDependencyClasspath ex: .jar files
    try {
      val skip = skipKey.value
      val libs = (dependencyClasspath in Runtime).value
      val projectJars = projectJarsKey.value
      val baseDir = baseDirectory.value
      val targetDir = target.value
      val jarOutputDir = crossTarget.value
      val copyFileNames = copiesKey.value
      if (!skip) doPackage(libs, projectJars, baseDir, targetDir, jarOutputDir, copyFileNames)
    } catch {
      case e: Exception =>
        println("xitrumPackage failed, reason:")
        e.printStackTrace()
    }
  }

  lazy val xitrumPackageNeedsPackageBin = (
    xitrumPackageKey := xitrumPackageKey.dependsOn(packageBin in Compile).value
  )

  //----------------------------------------------------------------------------

  private def doPackage(
    libs:          Seq[Attributed[File]],
    projectJars:   Seq[File],
    baseDir:       File,
    targetDir:     File,
    jarOutputDir:  File,
    copyFileNames: Seq[String]
  ) {
    val packageDir = targetDir / "xitrum"
    deleteFileOrDirectory(packageDir)
    packageDir.mkdirs()

    // Copy dependencies to lib directory
    val libDir = packageDir / "lib"
    libs.foreach { lib =>
      val file = lib.data

      if (file.exists) {
        if (file.isDirectory) {
          // This dependency may be "classes" directory from SBT multimodule (multiproject)
          // http://www.scala-sbt.org/0.13.5/docs/Getting-Started/Multi-Project.html
          //
          // Ex:
          // /Users/ngoc/src/xitrum-multimodule-demo/module1/target/scala-2.10/classes
          // /Users/ngoc/src/xitrum-multimodule-demo/module1/target/scala-2.10/xitrum-multimodule-demo-module1_2.10-1.0-SNAPSHOT.jar
          if (file.name == "classes") {
            val upperDir = file / ".."
            (upperDir * "*.jar").get.foreach { f =>
              val fname = f.name
              if (!fname.endsWith("-sources.jar") && !fname.endsWith("-javadoc.jar"))
                IO.copyFile(f, libDir / f.name)
            }
          }
        } else {
          IO.copyFile(file, libDir / file.name)
        }
      }
    }

    // Copy .jar files created after running "sbt package" to lib directory
    // (see xitrumPackageNeedsPackageBin)
    (jarOutputDir * "*.jar").get.foreach { f => IO.copyFile(f, libDir / f.name) }

    // Copy .jar files of dependency modules/projects
    projectJars.foreach { f => IO.copyFile(f, libDir / f.name) }

    copyFileNames.foreach { fName => doCopy(fName, baseDir, packageDir) }

    println("Packaged to " + packageDir)
  }

  private def deleteFileOrDirectory(file: File) {
    if (file.isDirectory) {
      val files = file.listFiles
      if (files != null) files.foreach { f => deleteFileOrDirectory(f) }
    }
    file.delete()
  }

  private def doCopy(fileName: String, baseDir: File, packageDir: File) {
    val file = new File(fileName)
    val from = if (file.isAbsolute) file else baseDir / fileName
    if (!from.exists) return

    val to = packageDir / file.name
    CopyUtil.copy(from, to)
  }

  // https://github.com/xerial/sbt-pack/blob/master/src/main/scala/xerial/sbt/pack/PackPlugin.scala

  val projectJarsKey = TaskKey[Seq[File]]("xitrumProjectJars")

  lazy val projectJarsTask = projectJarsKey := {
    Def.taskDyn {
      val libJars = getFromSelectedProjects(thisProjectRef.value, packageBin in Runtime, state.value)
      Def.task { libJars.value.map(_._1) }
    }.value
  }

  private def getFromSelectedProjects[T](contextProject: ProjectRef, targetTask: TaskKey[T], state: State): Task[Seq[(T, ProjectRef)]] = {
    val extracted = Project.extract(state)
    val structure = extracted.structure

    def transitiveDependencies(currentProject: ProjectRef): Seq[ProjectRef] = {
      // Traverse all dependent projects
      val children = Project
              .getProject(currentProject, structure)
              .toSeq
              .flatMap{ _.dependencies.map(_.project) }

      currentProject +: (children flatMap (transitiveDependencies(_)))
    }
    val projects = transitiveDependencies(contextProject).distinct
    projects.map(p => (Def.task { ((targetTask in p).value, p) }) evaluate structure.data).join
  }
}
